### PR TITLE
Product release 1/shared viewmodel

### LINKED ISSUE #1886: Edit Product Release 1: Implement shared viewModel across all edit product screens
Since all edit product related screens are linked, it would be better to use a shared viewmodel between all the product related screens. This would require migrating the Product detail screens to its own navigation sub graph.

### PR DESCRIPTION
**Pull request by anitaa1990**
Fixes #1886. This PR adds logic to migrate the product detail screens to its own navigation sub graph. Also completes the last step of #1614 by adding the logic to update product inventory details to the backend.

#### Changes:
- Modified the name of `ViewState` in `ProductDetailViewModel` to differentiate between the different `ViewStates` for multiple screens.
- Migrated product related screens to its own navigation sub graph.
- Created a `BaseProductFragment` to consolidate discard, menu button and exit related changes for product editing screens.
- When initialising a new screen, need to setup the `productFieldType`. Based on this field, the changes for this particular screen is discarded when discard button is clicked.
- As @ondrej mentioned, it might be a good idea to keep the `Product` model as read only and update fields using `copy()` function. Since this PR is already too big, I thought I would tackle this in another PR.

#### Known issues:
~- Currently when the Product inventory screen is rotated, the app crashes. I am still working on fixing this issue but this PR can be reviewed in the meantime.~ Fixed in 8be00ae

#### Testing:
- Since this PR migrates the product detail to it's own sub graph, it would be good to test clicking on product detail from the `Stats` and `Order` page as well.
- Verify that all information is displayed correctly in product detail page.
- You are able to update product title, description and inventory fields in the product detail page.

Update release notes:

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

**Comment by anitaa1990** [comment]
Sorry about a lot of noise on the PR! Was fixing a couple of issues and conflicts and failed tests. Done with everything and PR is ready for review 👍

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductInventoryFragment.kt [hunk 1]
```diff
@@ -82,39 +62,39 @@ class ProductInventoryFragment : BaseFragment(), ProductInventorySelectorDialogL
...
-        viewModel.viewStateLiveData.observe(viewLifecycleOwner) { old, new ->
-            new.productInventoryParameters?.takeIfNotEqualTo(old?.productInventoryParameters) { showProduct(new) }
-            new.isProductUpdated.takeIfNotEqualTo(old?.isProductUpdated) { showUpdateProductAction(it) }
-            new.skuErrorMessage?.takeIfNotEqualTo(old?.skuErrorMessage) { displaySkuError(it) }
+    override fun onOptionsItemSelected(item: MenuItem): Boolean {
+        return when (item.itemId) {
+            R.id.menu_done -> {
+                // TODO: add track event for click
+                ActivityUtils.hideKeyboard(activity)
+                viewModel.redirectToProductDetailScreen()
```
Comment (0nko) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductInventoryFragment.kt]: I don't think the fragment should tell the ViewModel what to do here. It should merely notify the VM that the done button was pressed and let it handle the logic.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductDetailViewModel.kt [hunk 2]
```diff
@@ -46,21 +49,39 @@ class ProductDetailViewModel @AssistedInject constructor(
...
+    }
+
     private var remoteProductId = 0L
     private var parameters: Parameters? = null
 
-    final val viewStateData = LiveDataDelegate(savedState, ViewState())
-    private var viewState by viewStateData
+    private var skuVerificationJob: Job? = null
+
+    final val commonStateLiveData = LiveDataDelegate(savedState, CommonViewState())
```
Comment (0nko) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductDetailViewModel.kt]: I noticed this is not being observed by any view other than a test. Is this planned for a future use?

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductDetailViewModel.kt [hunk 3]
```diff
@@ -79,6 +100,15 @@ class ProductDetailViewModel @AssistedInject constructor(
         triggerEvent(ShowImageChooser)
     }
 
+    fun redirectToProductDetailScreen() {
+        commonState = commonState.copy(shouldShowDiscardDialog = false)
+        triggerEvent(Exit)
+
+        // enable discard dialog once exit is triggered
+        commonState = commonState.copy(shouldShowDiscardDialog = true)
+        commonState = commonState.copy(productFieldType = null)
```
Comment (0nko) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductDetailViewModel.kt]: Could the state be updated in a single copy() call?

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductDetailViewModel.kt [hunk 4]
```diff
@@ -79,6 +100,15 @@ class ProductDetailViewModel @AssistedInject constructor(
         triggerEvent(ShowImageChooser)
     }
 
+    fun redirectToProductDetailScreen() {
+        commonState = commonState.copy(shouldShowDiscardDialog = false)
+        triggerEvent(Exit)
```
Comment (0nko) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductDetailViewModel.kt]: Instead of using the common `Exit` event you could create a new one, for example `ExitInventory`. There would be an inventory-specific observer/handler and you wouldn't need to check and update the `shouldShowDiscardDialog` flag this way. Just an idea, definitely not a blocker :).

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductInventoryFragment.kt [hunk 5]
```diff
@@ -125,21 +105,21 @@ class ProductInventoryFragment : BaseFragment(), ProductInventorySelectorDialogL
             isChecked = manageStock
             setOnCheckedChangeListener { _, b ->
                 enableManageStockStatus(b)
-                viewModel.updateProductInventoryDraft(manageStock = b)
+                viewModel.updateProductDraft(manageStock = b)
```
Comment (0nko) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductInventoryFragment.kt]: Nitpick: I'd use a more descriptive variable name for better readability here.

**Comment by anitaa1990** [comment]
Thanks for the review @0nko!

> State not reset
Go to inventory in product detail
Change the stock status
Tap the back button
Confirm to discard the changes
Go back to the inventory
Notice the previous selection is filled in

I'm not able to reproduce this 😢Do you mean even after clicking discard the `stockStatus` is showing the updated value? 




>Dialog disappearing

Yup, this is a known issue and we seem to handle different dialogs in the app in different ways. Some are dismissed when screen is rotated and some are retained. In order to keep this behaviour consistent across the apps, I have opened an issue [here](https://github.com/woocommerce/woocommerce-android/issues/1869) for this to tackle separately. 

> 1. Go to the inventory screen and fill out some data (for example stock quantity)
Tap the Done button to save it
Notice the change is reflected in the product detail screen
Go back to the inventory screen
Tap the back
A discard dialog is shown (even when no changes were made)
Tapping the Discard button resets the saved inventory screen state 

I was under the impression that this was the intended behaviour :) Since we have not yet saved the changes remotely, even if the user did not edit the screen now, it indicates that there are still some changes on this screen that are not yet saved remotely. So if they would like to discard those changes for some reason, (after clicking on back button and coming back to the inventory screen), this would be the only way to do that.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductDetailViewModel.kt (hunk 4 above)
Comment (anitaa1990) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductDetailViewModel.kt]: Good idea! Since all the Exit logic for all the screens are the same (i.e. we display discard dialog if there are some changes, otherwise just redirect to the previous screen), I didn't want to create a new Observer for each screen (even though this is a much cleaner approach😍) since in each screen we would need to call `activity.onBackPressed` like [here](https://github.com/woocommerce/woocommerce-android/blob/0e37dcc7901859cf0e13209f5976d854adbe9ddd/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/BaseProductFragment.kt#L60), which would need to check if discard dialog should be displayed or not and then redirect, which seemed more complicated. 

I'm in the middle of a brain freeze though so this is not what you meant I apologise.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductInventoryFragment.kt (hunk 5 above)
Comment (anitaa1990) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductInventoryFragment.kt]: Fixed in 8145875

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductDetailViewModel.kt (hunk 3 above)
Comment (anitaa1990) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductDetailViewModel.kt]: Fixed in 8145875

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductInventoryFragment.kt (hunk 1 above)
Comment (anitaa1990) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductInventoryFragment.kt]: Fixed in 947fcbe

**Comment by 0nko** [comment]
> I'm not able to reproduce this 😢Do you mean even after clicking discard the stockStatus is showing the updated value?

Yes. I'd expect any changes to be reset, since I discarded them. 

> I was under the impression that this was the intended behaviour :) Since we have not yet saved the changes remotely, even if the user did not edit the screen now, it indicates that there are still some changes on this screen that are not yet saved remotely. 

If that's the case, does it make sense to show the discard dialog after leaving a sub-product detail screen (inventory, description)? I mean, after clicking the Done button I'd assume the data had been saved. But after tapping the back button afterwards in the Product detail screen, the discard dialog is displayed again. 🤔

To me as a user, having the persistence logic done in one of 2 ways would be more clear:
1. Get rid of the discard dialogs and only keep a single one in the product detail.
2. Persist changes immediately upon leaving a subscreen.

Basically, showing the discard dialog indicates that I'm at a point where I need to either save the changes or discard them before leaving. Sorry for picking on details here, maybe I'm off and that's how it was intended by the design. If that's the case then feel free to disregard :).

**Comment by nbradbury** [comment]
> after clicking the Done button I'd assume the data had been saved .... Basically, showing the discard dialog indicates that I'm at a point where I need to either save the changes or discard them before leaving.

I think that's a very good point, and I'm also not sure our current approach will make the most sense to our users. Like @0nko, I feel like the user would expect changes to be saved as soon as they click the "Done" button on a sub-detail screen.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductDetailViewModel.kt (hunk 4 above)
Comment (anitaa1990) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductDetailViewModel.kt]: Thanks for the review @0nko!! I've made modifications as you mentioned in 9740987

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductDetailViewModel.kt (hunk 2 above)
Comment (anitaa1990) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductDetailViewModel.kt]: Good point! Removed/modified the logic in 9740987

**Comment by anitaa1990** [comment]
> I'm not able to reproduce this 😢Do you mean even after clicking discard the stockStatus is showing the updated value? 
Yes. I'd expect any changes to be reset, since I discarded them.

I'm not able to reproduce this @0nko 😢. I tried updating all fields in the inventory screen and then discarding them before coming back to the same screen. After discard, I am still seeing the  previous value before the update. This is the flow that I tested. Is there anything missing that can help with reproducing the steps?



>  after clicking the Done button I'd assume the data had been saved. But after tapping the back button afterwards in the Product detail screen, the discard dialog is displayed again.
I think that's a very good point, and I'm also not sure our current approach will make the most sense to our users. Like @0nko, I feel like the user would expect changes to be saved as soon as they click the "Done" button on a sub-detail screen.

Thanks for the clarification @0nko and @nbradbury! I always had the impression that the info wasn't actually saved until it was saved remotely and till then they might need to discard changes to a particular screen 🤔 I would like to make changes to this in a separate PR if that is fine? (This one is quite large as is). WDYT? 

Addressed all the other issues mentioned @0nko! Ready for another round 👍

**Comment by nbradbury** [comment]
> I would like to make changes to this in a separate PR if that is fine?

That's fine with me. I think we'll also want designer input on this, since it's a change from the original design (I think?).
